### PR TITLE
Collection improvements

### PR DESCRIPTION
**Pull request by krzysztof-gzocha**
Introduced:
- [x] Common `AbstractCollection`
- [x] Introduced `CellCollection`
- [x] Code improvements in `ChainSearch`
- [x] Documentation changes

Aiming to resolve: #90 and #92 

Humbug tests:
```
57 mutations were generated:
      57 mutants were killed
       0 mutants were not covered by tests
       0 covered mutants were not detected
       0 fatal errors were encountered
       0 time outs were encountered

Metrics:
    Mutation Score Indicator (MSI): 100%
    Mutation Code Coverage: 100%
    Covered Code MSI: 100%
```

### CODE REVIEW

File: src/KGzocha/Searcher/AbstractCollection.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
+<?php
+
+namespace KGzocha\Searcher;
+
+/**
+ * @author Krzysztof Gzocha <krzysztof@propertyfinder.ae>
+ */
+abstract class AbstractCollection implements \Countable, \IteratorAggregate
```
Comment (medinae) [inline src/KGzocha/Searcher/AbstractCollection.php]: 👍

**Comment by medinae** [comment]
AbstractCollection, CellCollection, modification in the Chain process, backward-compatible, full unit test coverage, humbug OK, Scrutinizer ok... seems to be excellent for me.

@drgomesp Can you also review ? Or any other interested dev ! This package definitely deserves a big community !

**Comment by krzysztof-gzocha** [comment]
Thanks @medinae for code review. 
Unfortunately it is not backward-compatible, because of changed constructor of `ChainProcess` and removed method `Cell::getName` - that's why I will merge this into `master-4.0` branch, not `master`. Rest is backward compatible.

**Comment by medinae** [comment]
True, my bad.

File: src/KGzocha/Searcher/Chain/Collection/CellCollection.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        return $item instanceof CellInterface;
+    }
+
+    /**
+     * @return \ArrayIterator
+     */
+    public function getIterator()
+    {
+        $this->validateNumberOfCells();
```
Comment (drgomesp) [inline src/KGzocha/Searcher/Chain/Collection/CellCollection.php]: Why is this here and not somewhere like in a constructor?

Comment (krzysztof-gzocha) [inline src/KGzocha/Searcher/Chain/Collection/CellCollection.php]: I would like to have possibility to create empty collection and then make use of `addNamedCell` or `addCell`, but I need to be sure there will be at least `MINIMUM_CELLS` when using this collection

**Comment by krzysztof-gzocha** [comment]
@drgomesp do you have any more comments?
